### ISSUE TITLE
zip with star unpack seems to have the wrong type.

### ISSUE THREAD

**Issue by AiroPi** [microsoft/pyright#4514]
**Describe the bug**
Use unpacking with zip (*zip()) in function argument seems to confuse pyright.

**To Reproduce**
Simply test with this simple code :
https://pyright-playground.decorator-factory.su/?gzip=H4sIAOMTz2MC_02OvQ7CMAyEdz-F1SlGAanlZ6gEEs_AiDq0IUGR2qSkZihPjxsBwpP1-Xx3LsUBTex7a9jHMG3azqAfxpgYL_bxtMFYcIuI59GH-_d2DjMA24lLPKIq5KvQWFZbyrDK8GadwN3-QACyo1Nt_TO9Tpwajd0f8YEbwvVpMa8BZZLlZwrYig7AqdXLjyqHaswxRBmrTwHSqKSDJuFvg4XDgtkAAAA%3D

**Expected behavior**
PyRight should not complains about typing

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
from collections.abc import Sequence
from typing import Any

test1 = ("abc", 123)
test2 = ("def", 456)

def f(a: Sequence[str], b: Sequence[int]) -> Any:
    return a, b

f(*zip(test1, test2))  # here it complains

f(*(("abc",), (123,)))  # here it doesn't, while technically it is similar.
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Both

**Comment by erictraut** [comment]
Pyright is working as expected here.

The type of the expression `zip(test1, test2)` is `zip[tuple[Literal["abc", 123], Literal["def", 456]]]`. When applying an unpack operator to this expression results in the type `tuple[Literal["abc", 123], Literal["def", 456]]`. This isn't compatible with parameters of type `Sequence[str]` and `Sequence[int]`. If you change the types of the two parameters to `Sequence[str | int]` it will type check correctly.

As evidence that pyright is doing the right thing here, mypy generates the same error.

**Comment by AiroPi** [comment]
But, if we do something like :

```py
test1 = ("abc", 123)
test2 = ("def", 456)

a = tuple(zip(test1, test2))
```
We will have `a = (('abc', 'def'), (123, 456))`.
That have a type `tuple[tuple[Literal['abc', 'def']], tuple[Literal[123, 456]]]`

And by using unpacking like `a = [*zip(test1, test2)]`, type is `list[tuple[Literal['abc', 123], Literal['def', 456]]]` while in reality, `a` is `[('abc', 'def'), (123, 456)]` so `list[tuple[Literal['abc', 'def']], tuple[Literal[123, 456]]]`.

**Comment by erictraut** [comment]
The `zip` function is defined in `builtins.pyi` (which comes from typeshed). Here's the relevant overload:

```python
class zip(Iterator[_T_co], Generic[_T_co]):
    ...
    @overload
    def __new__(cls, __iter1: Iterable[_T1], __iter2: Iterable[_T2], *, strict: bool = ...) -> zip[tuple[_T1, _T2]]: ...
    ...
```

In your code sample, `test1` is of type `Iterable[Literal['abc', 123]]`, and `test2` is of type `Iterable[Literal['def', 456]]`, so the `zip` constructor call results in a type `zip[tuple[Literal['abc', 123], Literal['def', 456]]`.

If you think that the `zip` type definition is incorrect, please report the problem to the maintainers of typeshed.
